### ISSUE TITLE
Test for false positives above PL-1 using the corpus

### ISSUE THREAD

**Issue by EsadCetiner** [coreruleset/coreruleset#4660]
### Description

PR https://github.com/coreruleset/coreruleset/pull/4476 was intended to resolve false positives against French addresses, but unintentionally introduced a regression that pretty much broke everything above PL-1. 942200 had only 3 tests before the PR was merged and no negative tests, it was only a matter of time until a PR would unintentionally break the rule.

I know generally false positives are more accepted at higher paranoia levels, but we shouldn't be tolerating a very sudden and huge spike in false positives. We should consider setting a threshold like 5% at PL-2, and 10% at PL-3 to try and catch these big increases. If this testing was in place for higher paranoia levels, it would've caught the regression and given us a chance to fix it before it ended up in a release:

v4.23.0 (Before the change): Total False positive ratio: 1252/100000 = 0.0125
v4.24.0 (After the change): Total False positive ratio: 100000/100000 = 1.0000

### How to reproduce the misbehavior (-> curl call)

```bash
N/A
```

### Logs

N/A

### Your Environment

N/A

### Confirmation

- [x] I have removed any personal data (email addresses, IP addresses, passwords, domain names) from any logs posted.

**Comment by fzipi** [comment]
CC @M4tteoP what do you think, is there an wasy way to warn abou this?

**Comment by fzipi** [comment]
## baseline false-positive rates per paranoia level

i measured the current false-positive rates on clean `main` (commit `8bfd6991`) using the same corpus the workflow uses (`eng` / `2023` / `10K`), running `ftw quantitative` separately at each paranoia level.

| PL | FP hits | corpus | aggregate ratio |
|----|--------:|-------:|----------------:|
| 1  | 84      | 10,000 | **0.84 %** |
| 2  | 717     | 10,000 | **7.17 %** |
| 3  | 4,606   | 10,000 | **46.06 %** |
| 4  | 17,210  | 10,000 | **172.1 %** |

three findings that change how we should set thresholds:

### 1. `falsePositives` counts rule *hits*, not flagged sentences

PL4's 17,210 exceeds the 10,000-sentence corpus because a single sentence can trip multiple rules. the per-PL bands are additive and consistent:

```
84 (PL1) + 633 (PL2) + 3,889 (PL3) + 12,604 (PL4) = 17,210
```

so an "FP ratio" above 100 % is normal, and the metric is not a clean per-sentence rate. any threshold wording needs to account for this.

### 2. the proposed thresholds (5 % PL2 / 10 % PL3) would fail immediately on clean `main`

current main is already at 7.17 % at PL2 and 46 % at PL3. the PL3 proposal is off by ~5×. fixed aggregate ceilings can't be picked from intuition — they have to be anchored to these measured baselines (e.g. PL2 ceiling ~10–12 %, PL3 ~55 %, with margin), and re-checked whenever the baseline shifts.

### 3. the aggregate is dominated by a few by-design-aggressive rules

top FP contributors at PL4:

- `920273` (PL4 strict charset): **10,000** — fires on essentially every natural-language sentence. expected.
- `920272` (PL3 strict charset): **3,489**. expected.
- `942432` (PL4 SQLi restricted chars): **2,604**. expected.
- `942200` (the rule from #4476): currently **199** at PL2 — recovered.

the #4476 regression was *one rule* blowing up, not the aggregate drifting. the JSON output already exposes `falsePositivesPerRule`, so a **per-rule new-vs-old delta gate** (flag any rule whose FP hits jump by more than N or X%) would catch a 942200-style spike precisely. an aggregate ratio gate at PL3/PL4 is permanently saturated by `920273` / `920272` and would either be numb or constantly red.

### recommendation

- extend the matrix to PL2 (high value, low noise) and PL3.
- gate on **per-rule** new-vs-old deltas rather than an aggregate ratio.
- don't gate the PL4 aggregate — it's saturated by intentional charset strictness; per-rule deltas still work there if we want PL4 coverage.

one GHA wrinkle to handle when extending the matrix: `manual_approval` currently gates on `needs.regression.outputs.test_status`, but matrix jobs don't aggregate outputs cleanly. switching that to `if: needs.regression.result != 'success'` (the job `result`, which does aggregate across the matrix) fixes it.

**Comment by fzipi** [comment]
Added enhancements to go-ftw as issues to have better quality in the quantitative runs.

**Comment by M4tteoP** [comment]
> CC @M4tteoP what do you think, is there an wasy way to warn abou this?

So far, the github action is just checking for increases in FPs at paranoia lv `1`, so even without adding thresholds, having the message reporting increases across all PLs would help. That being said, it looks like you are embedding what the GitHub action is doing in go-ftw, and it sounds like a great plan!

**Comment by fzipi** [comment]
@M4tteoP Do you think we need something else here, or can we close?

**Comment by M4tteoP** [comment]
The issue is mentioning warning based on ratios, I don't think this is implemented, but anyways, reporting the absolute numbers for all the PLs IMHO is good enough and catches any PL fps increase.

**Comment by M4tteoP** [comment]
Related to this,  https://github.com/coreruleset/coreruleset/pull/4772 is ready to be reviewed

**Comment by EsadCetiner** [comment]
My main concern was that we are completely blind to PRs that introduce false positive regressions that massively broke a rule, however that's no longer the case since the number of FPs per paranoia level is outputted.
